pkg/promtail/api/entry_parser_test.go

```diff
@@ -18,59 +20,79 @@ type Entry struct {
 	Labels model.LabelSet
 }
 
-func NewEntry(message string, stream string) Entry {
-	return Entry{TestTime, message, model.LabelSet{"stream": model.LabelValue(stream)}}
+func NewEntry(time time.Time, message string, stream string) Entry {
+	return Entry{time, message, model.LabelSet{"stream": model.LabelValue(stream)}}
 }
 
-var criTestCases = []struct {
+type TestCase struct {
 	Line     string // input
 	Error    bool
 	Expected Entry
-}{
+}
+
+var criTestCases = []TestCase{
 	{"", true, Entry{}},
 	{TestTimeStr, true, Entry{}},
 	{TestTimeStr + " stdout", true, Entry{}},
 	{TestTimeStr + " stderr F", true, Entry{}},
 	{TestTimeStr + " invalid F message", true, Entry{}},
 	{"2019-01-01 01:00:00.000000001 stdout F message", true, Entry{}},
 	{" " + TestTimeStr + " stdout F message", true, Entry{}},
-	{TestTimeStr + " stdout F message", false, NewEntry("message", "stdout")},
-	{TestTimeStr + " stderr P message", false, NewEntry("message", "stderr")},
-	{TestTimeStr + " stderr P message1\nmessage2", false, NewEntry("message1\nmessage2", "stderr")},
+	{TestTimeStr + " stdout F message", false, NewEntry(TestTime, "message", "stdout")},
+	{TestTimeStr + " stderr P message", false, NewEntry(TestTime, "message", "stderr")},
+	{TestTimeStr + " stderr P message1\nmessage2", false, NewEntry(TestTime, "message1\nmessage2", "stderr")},
 }
 
 func TestCRI(t *testing.T) {
-	for _, tc := range criTestCases {
+	runTestCases(CRI, criTestCases, t)
+}
+
+var dockerTestCases = []TestCase{
+	{
+		Line:     "{{\"log\":\"bad json, should fail to parse\\n\",\"stream\":\"stderr\",\"time\":\"2019-03-04T21:37:44.789508817Z\"}",
+		Error:    true,
+		Expected: Entry{},
+	},
+	{
+		Line:  "{\"log\":\"some silly log message\\n\",\"stream\":\"stderr\",\"time\":\"2019-03-04T21:37:44.789508817Z\"}",
+		Error: false,
+		Expected: NewEntry(time.Date(2019, 03, 04, 21, 37, 44, 789508817, time.UTC),
+			"some silly log message\n",
+			"stderr"),
+	},
+	{
+		Line:  "{\"log\":\"10.15.0.5 - - [04/Mar/2019:21:37:44 +0000] \\\"POST /api/prom/push HTTP/1.1\\\" 200 0 \\\"\\\" \\\"Go-http-client/1.1\\\"\\n\",\"stream\":\"stdout\",\"time\":\"2019-03-04T21:37:44.790195228Z\"}",
+		Error: false,
+		Expected: NewEntry(time.Date(2019, 03, 04, 21, 37, 44, 790195228, time.UTC),
+			"10.15.0.5 - - [04/Mar/2019:21:37:44 +0000] \"POST /api/prom/push HTTP/1.1\" 200 0 \"\" \"Go-http-client/1.1\"\n",
+			"stdout"),
+	},
+}
+
+func TestDocker(t *testing.T) {
+	runTestCases(Docker, dockerTestCases, t)
+}
+
+func runTestCases(parser EntryParser, testCases []TestCase, t *testing.T) {
+	for i, tc := range testCases {
 		client := &TestClient{
 			Entries: make([]Entry, 0),
 		}
 
-		EntryParser := CRI.Wrap(client)
-		err := EntryParser.Handle(model.LabelSet{}, time.Now(), tc.Line)
+		handler := parser.Wrap(client)
+		err := handler.Handle(model.LabelSet{}, time.Now(), tc.Line)
 		hasError := err != nil
 
 		if tc.Error != hasError {
```

Comment:
`if tc.Error == (err == nil) {`?

Reply:
I went a slightly different route with this, let me know if it works ok for you